# DATA WRANGLING PROJECT 
---------------------------------------
## WeRateDogs Twitter Data

#### 1. Importing Libraries

In [86]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import json
import requests
import tweepy
import os
import datetime, time

#### 2. Gathering Data

In [87]:
#get the twitter-archive-enhanced data
archive = pd.read_csv('twitter-archive-enhanced.csv')
archive.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   
2                  NaN                       NaN                        NaN   
3                  NaN                       NaN                        NaN   
4                  NaN                       NaN                        NaN   

                                                                                                                     expanded_urls  \
0                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0                13                  10   Phineas  None    None   None  None  
1                13                  10     Tilly  None    None   None  None  
2                12                  10    Archie  None    None   None  None  
3                13                  10     Darla  None    None   None  None  
4                12                  10  Franklin  None    None   None  None

In [3]:
#using the request library to get the image_predictions.tsv file programatically
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

response = requests.get(url)
open('image_predictions.tsv', 'wb').write(response.content)

335079

In [3]:
# load image_prediction.tsv
prediction = pd.read_csv('image_predictions.tsv', sep = "\t")
prediction.head(5)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [5]:
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_token_secret = 'HIDDEN'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

In [6]:
# get tweet status
tweet = api.get_status(archive.tweet_id[2], tweet_mode='extended')

Unauthorized: 401 Unauthorized
89 - Invalid or expired token.

In [ ]:
# show tweet info
info = tweet._json
info

In [55]:
# create tweet_json.txt file if it is no already existing
file = 'tweet_json.txt'
if not os.path.isfile(file):
    open(file, 'w', encoding = 'UTF-8')

In [56]:
# create a tweet id list
tweet_ids = archive.tweet_id.values

In [57]:
errors_tweets = []

# get start time of query
print("Start time:", datetime.datetime.now().time())
start = time.time()

# write JSON to .txt file
with open('tweet_json.txt', 'w', encoding = 'UTF-8') as file:
    for tweet_id in tweet_ids:
        try:
            tweet = api.get_status(tweet_id, tweet_mode = 'extended')
            json.dump(tweet._json, file)
            file.write('\n')
        except Exception as e:
            print("Error in Tweet ID:", tweet_id, "Time:",  datetime.datetime.now().time())
            errors_tweets.append(tweet_id)

# get end time of query
end = time.time()
print("End time:", datetime.datetime.now().time())

# display runtime

Start time: 13:27:07.303540
Error in Tweet ID: 888202515573088257 Time: 13:27:21.652570
Error in Tweet ID: 873697596434513921 Time: 13:28:16.988866
Error in Tweet ID: 872668790621863937 Time: 13:28:21.196561
Error in Tweet ID: 872261713294495745 Time: 13:28:23.381990
Error in Tweet ID: 869988702071779329 Time: 13:28:33.983069
Error in Tweet ID: 866816280283807744 Time: 13:28:44.314762
Error in Tweet ID: 861769973181624320 Time: 13:29:00.465679
Error in Tweet ID: 856602993587888130 Time: 13:29:19.772681
Error in Tweet ID: 856330835276025856 Time: 13:29:21.919683
Error in Tweet ID: 851953902622658560 Time: 13:29:41.103683
Error in Tweet ID: 851861385021730816 Time: 13:29:41.919679
Error in Tweet ID: 845459076796616705 Time: 13:30:06.992682
Error in Tweet ID: 844704788403113984 Time: 13:30:11.203701
Error in Tweet ID: 842892208864923648 Time: 13:30:16.069687
Error in Tweet ID: 837366284874571778 Time: 13:30:51.636151
Error in Tweet ID: 837012587749474308 Time: 13:30:53.109278
Error in Twe

Rate limit reached. Sleeping for: 126


Error in Tweet ID: 754011816964026368 Time: 13:42:32.325297
Error in Tweet ID: 680055455951884288 Time: 13:52:31.968876


Rate limit reached. Sleeping for: 224


End time: 14:04:20.961679


In [58]:
# display error IDs
errors_tweets

[888202515573088257,
 873697596434513921,
 872668790621863937,
 872261713294495745,
 869988702071779329,
 866816280283807744,
 861769973181624320,
 856602993587888130,
 856330835276025856,
 851953902622658560,
 851861385021730816,
 845459076796616705,
 844704788403113984,
 842892208864923648,
 837366284874571778,
 837012587749474308,
 829374341691346946,
 827228250799742977,
 812747805718642688,
 802247111496568832,
 779123168116150273,
 775096608509886464,
 771004394259247104,
 770743923962707968,
 766864461642756096,
 759923798737051648,
 759566828574212096,
 754011816964026368,
 680055455951884288]

In [59]:
# extract data from tweet_json.txt file
df_tweets = []
with open('tweet_json.txt') as json_file:
    for line in json_file:
        data = json.loads(line)
        df_tweets.append({'tweet_id': data['id'],
                        'retweet_count': data['retweet_count'],
                        'favorite_count': data['favorite_count']})    

In [60]:
# create dataframe
df_api = pd.DataFrame(df_tweets, columns= ['tweet_id',
                                           'retweet_count',
                                           'favorite_count'])

In [61]:
df_api.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7010           33824
1  892177421306343426           5302           29337
2  891815181378084864           3482           22069
3  891689557279858688           7228           36950
4  891327558926688256           7764           35321

In [62]:
# save dataframe
df_api.to_csv('df_tweets.csv')

-------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [5]:
tweets_df = pd.read_csv('df_tweets.csv')

## Data Claening

#### Assessing Data: Quality and Tidiness Issues

> Visual assesment

In [88]:
# display complete row
pd.set_option('display.max_colwidth', None)
display(archive.head(10))
display(archive.tail(10))

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   
5  891087950875897856                    NaN                  NaN   
6  890971913173991426                    NaN                  NaN   
7  890729181411237888                    NaN                  NaN   
8  890609185150312448                    NaN                  NaN   
9  890240255349198849                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   
5  2017-07-29 00:08:17 +0000   
6  2017-07-28 16:27:12 +0000   
7  2017-07-28 00:22:40 +0000   
8  2017-07-27 16:25:51 +0000   
9  2017-07-26 15:59:51 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
5  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
6  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
7  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
8  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
9  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                     text  \
0                                                           This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1      This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                       This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                                 This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4      This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   
5      Here we have a majestic great white breaching off South Africa's coast. Absolutely h*ckin breathtaking. 13/10 (IG: tucker_marlo) #BarkWeek https://t.co/kQ04fDDRmh   
6  Meet Jax. He enjoys ice cream so much he gets nervous around it. 13/10 help Jax enjoy more things by clicking below\n\nhttps://t.co/Zr4hWfAs1H https://t.co/tVJBRMnhxl   
7                          When you watch your owner call another dog a good boy but then they turn back to you and say you're a great boy. 13/10 https://t.co/v0nONBcwxq   
8                      This is Zoey. She doesn't want to be one of the scary sharks. Just wants to be a snuggly pettable boatpet. 13/10 #BarkWeek https://t.co/9TwLuAGH0b   
9           This is Cassie. She is a college pup. Studying international doggo communication and stick theory. 14/10 so elegant much sophisticate https://t.co/t1bfwz5S2A   

   retweeted_sta

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN                  NaN   
2350  666050758794694657                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2346  2015-11-16 01:01:59 +0000   
2347  2015-11-16 00:55:59 +0000   
2348  2015-11-16 00:49:46 +0000   
2349  2015-11-16 00:35:11 +0000   
2350  2015-11-16 00:30:50 +0000   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                                                  source  \
2346  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2347  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2348  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2349  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2350  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2352  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2353  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2354  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2355  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
2346       Here is the Rand Paul of retrievers folks! He's probably good at poker. Can drink beer (lol rad). 8/10 good dog https://t.co/pYAJkAe76p   
2347                  My oh my. This is a rare blond Canadian terrier on wheels. Only $8.98. Rather docile. 9/10 very rare https://t.co/yWBqbrzy8O   
2348  Here is a Siberian heavily armored polar bear mix. Strong owner. 10/10 I would do unspeakable things to pet this dog https://t.co/rdivxLiqEt   
2349    This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc   
2350  This is a truly beautiful English Wilson Staff retriever. Has a nice phone. Privileged. 10/10 would trade lives with https://t.co/fvIbQfHjIe   
2351                      Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq   
2352     This is a purebred Piers Morgan. Loves to Netflix and chill. Always looks like he forgot to unplug the iron. 6/10 https://t.co/DWnyCjf2mx   
2353            Here is a very happy pup. Big fan of well-maintained decks. Just look at that tongue. 9/10 would cuddle af https://t.co/y671yMhoiR   
2354   This is a western brown Mitsubishi terrier. Upset about leaf. Actually 2 dogs here. 7/10 would walk the shit out of https://t.co/r7mOb2m0UI   
2355           Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj   

      retweeted_status_id  retweeted_status_user_id  \
2346                  NaN                       NaN   
2347                  NaN                       NaN   
2348    

> programatic assessment

In [89]:
archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [90]:
#checkign for unique tweet_id values
archive.tweet_id.nunique()

2356

In [91]:
#checking for null values
archive.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [92]:
# count numbers of values in columns doggo, floofer, pupper and puppo
(archive.loc[:, 'doggo':'puppo'] != 'None').sum()

doggo       97
floofer     10
pupper     257
puppo       30
dtype: int64

In [93]:
# check whether all dog names are filled
archive.name.value_counts()

None       745
a           55
Charlie     12
Oliver      11
Cooper      11
          ... 
such         1
Todo         1
Buckley      1
Ridley       1
Acro         1
Name: name, Length: 957, dtype: int64

In [94]:
# check whether rating_numerator is always filled
archive.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [95]:
archive.rating_numerator.describe()

count    2356.000000
mean       13.126486
std        45.876648
min         0.000000
25%        10.000000
50%        11.000000
75%        12.000000
max      1776.000000
Name: rating_numerator, dtype: float64

In [96]:
# check whether rating_demoninator is always filled
archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [97]:
archive.rating_denominator.describe()

count    2356.000000
mean       10.455433
std         6.745237
min         0.000000
25%        10.000000
50%        10.000000
75%        10.000000
max       170.000000
Name: rating_denominator, dtype: float64

In [98]:
# check sources
archive.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

###### PREDICTION TABLE

> Visual Assessment

In [99]:
display(prediction.head())
display(prediction.tail())

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

> Programatic Assessment

In [100]:
prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [101]:
#checking for null values
prediction.isnull().sum()

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

###### TWEET_DF TABLE

> Visual Assesment

In [102]:
display(tweets_df.head())
display(tweets_df.tail())

Unnamed: 0            tweet_id  retweet_count  favorite_count
0           0  892420643555336193           7010           33824
1           1  892177421306343426           5302           29337
2           2  891815181378084864           3482           22069
3           3  891689557279858688           7228           36950
4           4  891327558926688256           7764           35321

Unnamed: 0            tweet_id  retweet_count  favorite_count
2322        2322  666049248165822465             37              89
2323        2323  666044226329800704            115             247
2324        2324  666033412701032449             36             100
2325        2325  666029285002620928             39             112
2326        2326  666020888022790149            423            2292

> Programatic Assessment

In [103]:
tweets_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2327 entries, 0 to 2326
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   Unnamed: 0      2327 non-null   int64
 1   tweet_id        2327 non-null   int64
 2   retweet_count   2327 non-null   int64
 3   favorite_count  2327 non-null   int64
dtypes: int64(4)
memory usage: 72.8 KB


In [104]:
archive.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   

                                                     expanded_urls  \
0  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1  https://twitter.com/dog_rates/status/892177421306343426/photo/1   

   rating_numerator  rating_denominator     name doggo floofer pupper puppo  
0                13                  10  Phineas  None    None   None  None  
1                13                  10    Tilly  None    None   None  None

#### QUALITY
###### 'archive' Table
- There are 23 cases where the denominator of rating != 10. These entries will be removed. It is possible some of these entries are genuine tweets, in which case each of these entries would have to be inspected individually.
-  Multiple cases of where the numerator of rating < 10. These entries will be removed.
- Multiple cases where numerator of rating >> 10. These entries will be assessed and removed if not authentic.
- Rename columns with more appropriate names: "timestamp" to "tweet_timestamp", "text" to "tweet_text", "rating_numerator" to "dog_rating_out_of_ten", "name" to "dog_name"
-  Since retweets and replies will be removed, the column "retweeted_status_timestamp" will be removed as it will no longer provide any useful information
- Remove column "rating_denominator" once all the values that != 10 have been removed since this will no longer provide any useful information

###### 'prediction' Table
- Entries where the first (i.e. most confident prediction) has a False value for "p1_dog" (i.e. does not correspond to a type of dog) will be removed.
- The "p1" and "p1_conf" columns will be renamed with more explanatory titles.
- The column "jpg_url" will be removed since url data is already contained in the twitter archive data
- The "p2" and "p3" related columns will be removed as I am only using the most likely prediction ("p1") in my analysis
- After removal of "False" entries, the "p1_dog" column will be removed as it will no longer add any valuable information.

#### TIDINESS
###### 'archive' Table
- There are 181 retweets which need to be removed. This will ensure there are not two rows corresponding to the same tweet, i.e. holding to the definition of tidy data which requires each row to represent a unique entry.
- There are 78 tweet replies which need to be removed. This will ensure there are not two rows corresponding to the same tweet, i.e. holding to the definition of tidy data which requires each row to represent a unique entry.
- Change columns "doggo", "floofer", "pupper", and "puppo" from wide to long format.


### CLEANING DATA

In [105]:
archive_clean = archive.copy()

##### QUALITY ISSUES

###### Define
- Removing multiple cases of where the denominator of rating != 10.

###### Code

In [106]:

archive_clean.drop(archive_clean[archive_clean.rating_denominator != 10].index, inplace = True)

In [107]:
#Tets
archive_clean.rating_denominator.value_counts()

10    2333
Name: rating_denominator, dtype: int64

###### Define
-  Multiple cases of where the numerator of rating < 10. These entries will be removed..

###### Code

In [108]:
archive_clean.drop(archive_clean[archive_clean.rating_numerator < 10].index, inplace = True)

In [109]:
# Test
archive_clean.rating_numerator.value_counts()

12      558
11      463
10      461
13      351
14       54
420       2
75        2
15        2
1776      1
666       1
182       1
27        1
26        1
17        1
Name: rating_numerator, dtype: int64

 I have decided to carry out steps 1 and 2 in tidiness first and then return to c) in the quality as this will likely remove many of the values present in assessment c)

#### TIDINESS

###### Define
- There are 181 retweets which need to be removed.

###### Code

In [110]:
archive_clean.drop(archive_clean[archive_clean.retweeted_status_id.notnull()].index, inplace = True)

In [111]:
# Test
archive_clean.retweeted_status_id.notnull().sum()

0

###### Define
- There are 78 tweet replies which need to be removed. 

###### Code

In [112]:
archive_clean.drop(archive_clean[archive_clean.in_reply_to_status_id.notnull()].index, inplace = True)

In [113]:
# Test
archive_clean.in_reply_to_status_id.notnull().sum()

0

> Moving back to the quality issues

###### Define
- Multiple cases where numerator of rating >> 10. These entries will be assessed and removed if not authentic. 

###### Code 

In [114]:
archive_clean.rating_numerator.value_counts()

12      486
10      436
11      413
13      287
14       38
1776      1
420       1
75        1
27        1
26        1
Name: rating_numerator, dtype: int64

In [115]:
# There appear to be 5 cases where the numerator is exceptionally high (> 14)

index  = [420, 1776,75,27,26]

for x in index:
    print(f'Rating number: {x}')
    for i in range(0,len(archive.index)):
        if archive.rating_numerator[i] == x:
            print(archive.tweet_id[i])
            print(archive.text[i])
            print(archive.expanded_urls[i])
            print(archive.retweeted_status_id[i])
            print(archive.in_reply_to_status_id[i])
            print(archive.source[i])
            print(archive.index[i])
            print('\n')

        else:
            pass

Rating number: 420
855862651834028034
@dhmontgomery We also gave snoop dogg a 420/10 but I think that predated your research
nan
nan
8.558615844633518e+17
<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
188


670842764863651840
After so many requests... here you go.

Good dogg. 420/10 https://t.co/yfAAo1gdeY
https://twitter.com/dog_rates/status/670842764863651840/photo/1
nan
nan
<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>
2074


Rating number: 1776
749981277374128128
This is Atticus. He's quite simply America af. 1776/10 https://t.co/GRXwMxLBkh
https://twitter.com/dog_rates/status/749981277374128128/photo/1
nan
nan
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>
979


Rating number: 75
832215909146226688
RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…
https://twitter.com/dog_rates/statu

In [116]:
# All 5 cases appear to be genuine. However, the numerator value is incorrect for some and will be updated:

nums = [75,27,26]
correct_vals = [9.75, 11.27, 11.26]
index_list = []

# retrieving indexes
for i in nums:
    index_list.append(archive_clean[archive_clean.rating_numerator == i].index[0])

# Replacing with correct values
for i,j in zip(index_list,correct_vals):
    archive_clean.at[i, 'rating_numerator'] = j

In [117]:
# Test
archive_clean.rating_numerator.value_counts()

12      486
10      436
11      415
13      287
14       38
1776      1
420       1
9         1
Name: rating_numerator, dtype: int64

###### Define
- Rename columns with more appropriate names: "timestamp" to "tweet_timestamp", "text" to "tweet_text", "rating_numerator" to "dog_rating_out_of_ten", "name" to "dog_name"

###### Code

In [118]:
# Code
new_names = {'timestamp': 'tweet_timestamp', 'text': 'tweet_text',\
             'rating_numerator': 'dog_rating_out_of_ten', 'name': 'dog_name'}


archive_clean.rename(columns=new_names, inplace=True)

In [119]:
# Test
list(archive_clean.columns)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'tweet_timestamp',
 'source',
 'tweet_text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'dog_rating_out_of_ten',
 'rating_denominator',
 'dog_name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

###### Define

- Since retweets and replies will be removed, the column "retweeted_status_timestamp" will be removed as it will no longer provide any useful information.

- Remove column "rating_denominator" once all the values that != 10 have been removed since this will no longer provide any useful information.

###### Code


In [120]:
drop_cols = ['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id',\
             'retweeted_status_user_id','retweeted_status_timestamp','rating_denominator']

archive_clean.drop(drop_cols, inplace=True, axis=1)

In [121]:
# Test
list(archive_clean.columns)

['tweet_id',
 'tweet_timestamp',
 'source',
 'tweet_text',
 'expanded_urls',
 'dog_rating_out_of_ten',
 'dog_name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

##### TIDINESS ISSUE

###### Define
- Change columns "doggo", "floofer", "pupper", and "puppo" from wide to long format and make the datatype category.

###### Code

In [122]:
archive_clean = archive_clean.reset_index()
del archive_clean['index']

In [41]:
archive_clean['dog_category'] = 'NaN'

In [124]:
# Creating a single column for dog category

for i in range(0, len(archive_clean.index)):
    if archive_clean.loc[i, 'doggo'] == 'doggo':
        archive_clean.loc[i, 'dog_category'] = 'doggo'
    elif archive_clean.loc[i, 'floofer'] == 'floofer':
        archive_clean.loc[i, 'dog_category'] = 'floofer'
    elif archive_clean.loc[i, 'pupper'] == 'pupper':
        archive_clean.loc[i, 'dog_category'] = 'pupper'
    elif archive_clean.loc[i, 'puppo'] == 'puppo':
        archive_clean.loc[i, 'dog_category'] = 'puppo'
    else:
        archive_clean.loc[i, 'dog_category'] = 'None'

In [125]:
# Checking for entries labelled with multiple categories

multi_cat_list = []

for i in range(0, len(archive_clean.index)):
    if archive_clean.doggo[i] == 'doggo' and (archive_clean.floofer[i] == 'floofer' \
                                                   or archive_clean.pupper[i] == 'pupper'\
                                                   or archive_clean.puppo[i] == 'puppo'):
        multi_cat_list.append(i)

print(multi_cat_list)
print(len(multi_cat_list))

[153, 160, 354, 411, 441, 531, 556, 677, 813, 852]
10


>10 cases of multiple categories found 
>Checking text of entry to correct category to correct values

In [128]:
# Index 153    
archive_clean.loc[153, ['doggo', 'floofer', 'pupper', 'puppo', 'dog_category']]

doggo           doggo
floofer          None
pupper           None
puppo           puppo
dog_category    doggo
Name: 153, dtype: object

In [129]:
archive_clean.tweet_text[153]

"Here's a puppo participating in the #ScienceMarch. Cleverly disguising her own doggo agenda. 13/10 would keep the planet habitable for https://t.co/cMhq16isel"

In [130]:
# Incorreclty labelled as "doggo", correcting to "puppo"

archive_clean.loc[153, 'dog_category'] = 'puppo'

In [132]:
# Index 160
archive_clean.loc[160, ['doggo', 'floofer', 'pupper', 'puppo', 'dog_category']]

doggo             doggo
floofer         floofer
pupper             None
puppo              None
dog_category      doggo
Name: 160, dtype: object

In [134]:
archive_clean.tweet_text[160]

"At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk"

In [43]:
# create list of stages
dog_list =['doggo', 'floofer', 'pupper', 'puppo']

In [44]:
# create new column with empty lists
archive_clean['dog_type'] = pd.np.empty((archive_clean.shape[0], 0)).tolist()

<ipython-input-44-4cf1f0041988>:2: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead
  archive_clean['dog_type'] = pd.np.empty((archive_clean.shape[0], 0)).tolist()


In [45]:
# append list in cells of column 'dog_type' with values from other columns
def d_type(archive_clean):    
    for i in range(archive_clean.shape[0]):
        for x in dog_list:
            if x in archive_clean.loc[i,['doggo', 'floofer', 'pupper', 'puppo']].tolist():
                archive_clean.loc[i,'dog_type'].append(x)
            else:
                continue   
        # convert list into string
        archive_clean.loc[i,'dog_type'] = ", ".join(archive_clean.loc[i,'dog_type'])
    # replace empty strings with another string
    archive_clean.dog_type = archive_clean.dog_type.replace('',np.nan)
    #archive_clean.dog_type = archive_clean.dog_type.replace(np.nan)


In [46]:
d_type(archive_clean)

In [47]:
# Test
archive_clean.loc[(archive_clean[['doggo', 'floofer', 'pupper', 'puppo']] != 'None').sum(axis=1) > 1].head()

tweet_id            tweet_timestamp  \
153  855851453814013952  2017-04-22 18:31:02 +0000   
160  854010172552949760  2017-04-17 16:34:26 +0000   
354  817777686764523521  2017-01-07 16:59:28 +0000   
411  808106460588765185  2016-12-12 00:29:28 +0000   
441  801115127852503040  2016-11-22 17:28:25 +0000   

                                                                                 source  \
153  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
160  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
354  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
411  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
441  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                               tweet_text  \
153        Here's a puppo participating in the #ScienceMarch. Cleverly disguising her own doggo agenda. 13/10 would keep the planet habitable for https://t.co/cMhq16isel   
160  At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk   
354  This is Dido. She's playing the lead role in "Pupper Stops to Catch Snow Before Resuming Shadow Box with Dried Apple." 13/10 (IG: didodoggo) https://t.co/m7isZrOBX7   
411                      Here we have Burke (pupper) and Dexter (doggo). Pupper wants to be exactly like doggo. Both 12/10 would pet at same time https://t.co/ANBpEYHaho   
441                      This is Bones. He's being haunted by another doggo of roughly the same size. 12/10 deep breaths pupper everything's fine https://t.co/55Dqe0SJNj   

                                                                                                                       expanded_urls  \
153                                                                  https://twitter.com/dog_rates/status/855851453814013952/photo/1   
160  https://twitter.com/dog_rates/status/854010172552949760/photo/1,https://twitter.com/dog_rates/status/854010172552949760/photo/1   
354                                                                  https://twitter.com/dog_rates/status/817777686764523521/video/1   
411                                                                  https://twitter.com/dog_rates/status/808106460588765185/photo/1   
441  https://twitter.com/dog_rates/status/801115127852503040/photo/1,https://twitter.com/dog_rates/status/801115127852503040/photo/1   

     dog_rating_out_of_ten dog_name  doggo  floofer  pupper  puppo  \
153                     13     None  doggo     None    None  puppo   
160                     11     None  doggo  floofer    None   None   
354                     13     Dido  doggo     None  pupper   None   
411                     12     None  doggo     None  pupper   None   
441                     12    Bones  doggo     None  pupper   None   

    dog_category        dog_type  
153          NaN    doggo, puppo  
160          NaN  doggo, floofer  
354          NaN   doggo, pupper  
411          NaN   doggo, pupper  
441          NaN   doggo, pupper

In [48]:
# drop old columns
archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis = 1, inplace = True)

In [49]:
archive_clean.dog_type = archive_clean.dog_type.astype('category')

In [50]:
#check whether NaNs still in table
archive_clean.dog_type.head(15)

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
5       NaN
6       NaN
7       NaN
8       NaN
9     doggo
10      NaN
11      NaN
12    puppo
13      NaN
14    puppo
Name: dog_type, dtype: category
Categories (7, object): ['doggo', 'doggo, floofer', 'doggo, pupper', 'doggo, puppo', 'floofer', 'pupper', 'puppo']

In [51]:
#count dog types in new column
archive_clean.dog_type.value_counts()

pupper            182
doggo              69
puppo              22
floofer             9
doggo, pupper       8
doggo, puppo        1
doggo, floofer      1
Name: dog_type, dtype: int64

In [52]:
archive_clean.head()

tweet_id            tweet_timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                           tweet_text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   dog_rating_out_of_ten  dog_name dog_category dog_type  
0                     13   Phineas          NaN      NaN  
1                     13     Tilly          NaN      NaN  
2                     12    Archie          NaN      NaN  
3                     13     Darla          NaN      NaN  
4                     12  Franklin          NaN      NaN

##### PREDICTION DATA

In [55]:
prediction_clean= prediction.copy()

##### QUALITY ISSUE

###### Define

- Entries where the first (i.e. most confident prediction) has a False value for "p1_dog" (i.e. does not correspond to a type of dog) will be removed. 

###### Code

In [56]:
prediction_clean.drop(prediction_clean[prediction_clean.p1_dog == False].index, inplace=True)

In [58]:
# Test
prediction_clean.p1_dog.value_counts()

True    1532
Name: p1_dog, dtype: int64

###### Define
- The "p1" and "p1_conf" columns will be renamed with more explanatory titles.

###### Code

In [59]:
new_col_names = {'p1':'dog_breed_prediction', 'p1_conf':'prediction_confidence'}

prediction_clean.rename(columns= new_col_names, inplace=True)

In [61]:
# Test
list(prediction_clean.columns)

['tweet_id',
 'jpg_url',
 'img_num',
 'dog_breed_prediction',
 'prediction_confidence',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog']

###### Define

- The column "jpg_url" will be removed since url data is already contained in the twitter archive data (expanded_urls)

- The "p2" and "p3" related columns will be removed as I am only using the most likely prediction ("p1") in my analysis

- After removal of "False" entries, the "p1_dog" column will be removed as it will no longer add any valuable information.

###### Code

In [62]:
preds_drop_cols = ['jpg_url', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p2_conf', 'p3_dog', 'p3_conf']

prediction_clean.drop(preds_drop_cols, inplace=True, axis=1)

In [64]:
# Test
list(prediction_clean.columns)

['tweet_id', 'img_num', 'dog_breed_prediction', 'prediction_confidence']

#### COMBINING THE DATA

In [79]:
# Merging dataframes to create a master dataframe 

master_df_pt1 = pd.merge(archive_clean, prediction_clean, on='tweet_id', how='left')

In [80]:
master_df_pt1.head()

tweet_id            tweet_timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                           tweet_text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   dog_rating_out_of_ten  dog_name dog_category dog_type  img_num  \
0                     13   Phineas          NaN      NaN      NaN   
1                     13     Tilly          NaN      NaN      1.0   
2                     12    Archie          NaN      NaN      1.0   
3                     13     Darla          NaN      NaN      NaN   
4                     12  Franklin          NaN      NaN      2.0   

  dog_breed_prediction  prediction_confidence  
0                  NaN                    NaN  
1            Chihuahua               0.323581  
2            Chihuahua               0.716012  
3                  NaN                    NaN  
4               basset               0.555712

In [81]:
master_df = pd.merge(master_df_pt1, tweets_df_clean,on=['tweet_id'],how='left')

In [83]:
# Saving data to csv

master_df.to_csv('twitter_archive_master.csv')
archive_clean.to_csv('clean_twitter_archive_data.csv')
prediction_clean.to_csv('clean_image_prediction_data.csv')
tweets_df_clean.to_csv('clean_rt_and_fav_data.csv')

ANALYSIS

In [84]:
master_df.head()

tweet_id            tweet_timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                           tweet_text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

                                                                                                                     expanded_urls  \
0                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   dog_rating_out_of_ten  dog_name dog_category dog_type  img_num  \
0                     13   Phineas          NaN      NaN      NaN   
1                     13     Tilly          NaN      NaN      1.0   
2                     12    Archie          NaN      NaN      1.0   
3                     13     Darla          NaN      NaN      NaN   
4                     12  Franklin          NaN      NaN      2.0   

  dog_breed_prediction  prediction_confidence  Unnamed: 0  retweet_count  \
0                  NaN                    NaN         0.0         7010.0   
1            Chihuahua               0.323581         1.0         5302.0   
2            Chihuahua               0.716012         2.0         3482.0   
3                  NaN                    NaN         3.0         7228.0   
4               basset               0.555712         4.0         7764.0   

   favorite_count  
0         33824.0  
1         29337.0  
2         22069.0  
3         36950.0  
4         35321.0

In [85]:
master_df.dog_category.value_counts()

NaN    1665
Name: dog_category, dtype: int64